modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,821 @@
+  }
+}
+
+DlsoBackend::~DlsoBackend()
+{
+  if (this->api->release != nullptr) {
+    this->api->release(this->api);
+  }
+  dlclose(this->dlhandle);
+}
+
+void DlsoBackend::lookup(const QType& qtype, const DNSName& qdomain, int32_t zoneId, DNSPacket* pkt_p)
+{
+  DLOG(
+    g_log << Logger::Debug << __PRETTY_FUNCTION__ << " qtype: " << qtype.toString() << " qdomain: " << qdomain << " zoneId: " << zoneId << " pkt_p: " << pkt_p << endl;
+  );
+  if (in_query) {
+    throw PDNSException("Attempt to lookup while one running");
+  }
+
+  string qname = qdomain.toString();
+  bool success = false;
+
+  if (pkt_p != nullptr) {
+    ComboAddress edns_or_resolver_ip = pkt_p->getRealRemote().getNetwork();
+    // TODO expose the whole union to 3rd party lib.
+    // clang-tidy: This reinterpret_cast is needed as we expose a generic `sockaddr`.
+    auto* sockaddr = reinterpret_cast<struct sockaddr*>(std::addressof(edns_or_resolver_ip.sin4));
+    success = api->lookup(api->handle, qtype.getCode(), qname.size(), qname.c_str(), sockaddr, zoneId);
+  }
+  else {
+    success = api->lookup(api->handle, qtype.getCode(), qname.size(), qname.c_str(), nullptr, zoneId);
+  }
+
+  if (!success) {
+    throw PDNSException("Backend failed");
+  }
+}
+
+bool DlsoBackend::list(const DNSName& target, int domain_id, bool include_disabled)
+{
+  DLOG(
+    g_log << Logger::Debug << __PRETTY_FUNCTION__ << " target: " << target << " domain_id " << domain_id << endl;
+  );
+  if (api->list == nullptr) {
+    return false;
+  }
+
+  string qname = target.toString();
+  bool success = api->list(api->handle, qname.size(), qname.c_str(), domain_id);
+
+  if (!success) {
+    throw PDNSException("Backend failed");
+  }
+
+  return success;
+}
+
+extern "C" void fill_cb(void* ptr, const struct resource_record* record)
+{
+  auto* rr = static_cast<DNSResourceRecord*>(ptr);
+  rr->qtype = record->qtype;
+  string qname = string(record->qname, record->qname_len);
+  rr->qname = DNSName(qname);
+  rr->qclass = QClass::IN;
+  rr->content = string(record->content, record->content_len);
+  rr->ttl = record->ttl;
+  rr->auth = record->auth;
+  rr->scopeMask = record->scope_mask;
+  rr->domain_id = record->domain_id;
+  DLOG(
+    g_log << Logger::Debug << __PRETTY_FUNCTION__
+      << " DNSResourceRecord qtype: " << rr->qtype
+      << " qname: " << rr->qname
+      << " content: " << rr->content
+      << " ttl: " << rr->ttl
+      << " domain_id: " << rr->domain_id
+      << " auth: " << rr->auth << endl;
+  );
+}
+
+bool DlsoBackend::get(DNSResourceRecord& rr)
+{
+  DLOG(
+    g_log << Logger::Debug << __PRETTY_FUNCTION__ << endl;
+  );
+  bool success = api->get(api->handle, fill_cb, &rr);
+
+  if (!success) {
+    in_query = false;
+    return false;
+  }
+
+  return true;
+}
+
+extern "C" void fill_meta_cb(void* ptr, uint8_t value_len, const struct dns_value* values)
+{
+  auto* meta = static_cast<std::vector<std::string>*>(ptr);
+  meta->resize(meta->size() + value_len);
```

Comment:
```suggestion
  meta->reserve(meta->size() + value_len);
```

Hello, I've been working on setting things to enable pdns to accept notify host request from pdns control and send corresponding signed notifies and I stumbled onto this. It does not seem like this part was used somewhere else in our setup for a long time. 

With this change, pdns is able to found the needed key and send the notify.

However I'm not fluent in C++, so there may be a better solution.

Thanks !